pkg/webhook/external/list.go

```diff
@@ -0,0 +1,186 @@
+package external
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/pkg/expressions"
+)
+
+func (g *genericWebhookReceiver) listUniqueObjects(
+	ctx context.Context,
+	action kargoapi.GenericWebhookAction,
+	actionEnv map[string]any,
+) ([]client.Object, []error) {
+	var resources []client.Object
+	var listErrs []error
+	for i, tsc := range action.TargetSelectionCriteria {
+		objects, err := g.listTargetObjects(ctx, tsc, actionEnv)
+		if err != nil {
+			listErrs = append(listErrs,
+				fmt.Errorf("selection criteria error at index %d: %w", i, err),
+			)
+			continue
+		}
+		resources = append(resources, objects...)
+	}
+	slices.SortFunc(resources, func(a, b client.Object) int {
+		if comp := strings.Compare(a.GetNamespace(), b.GetNamespace()); comp != 0 {
+			return comp
+		}
+		return strings.Compare(a.GetName(), b.GetName())
+	})
+	resources = slices.CompactFunc(resources, func(a, b client.Object) bool {
+		return a.GetNamespace() == b.GetNamespace() && a.GetName() == b.GetName()
+	})
+	return resources, listErrs
+}
+
+func (g *genericWebhookReceiver) listTargetObjects(
+	ctx context.Context,
+	targetSelectionCriteria kargoapi.GenericWebhookTargetSelectionCriteria,
+	actionEnv map[string]any,
+) ([]client.Object, error) {
+	listOpts, err := g.buildListOptionsForTarget(targetSelectionCriteria, actionEnv)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build list options: %w", err)
+	}
+	switch targetSelectionCriteria.Kind {
+	case kargoapi.GenericWebhookTargetKindWarehouse:
+		warehouses := new(kargoapi.WarehouseList)
+		if err := g.client.List(ctx, warehouses, listOpts...); err != nil {
+			return nil, fmt.Errorf("error listing %s targets: %w", targetSelectionCriteria.Kind, err)
+		}
+		if targetSelectionCriteria.Name == "" {
+			return itemsToObjects(warehouses.Items), nil
+		}
+		name, err := evalAsString(targetSelectionCriteria.Name, actionEnv)
+		if err != nil {
+			return nil, fmt.Errorf("failed to evaluate target name as string: %w", err)
+		}
+		var filtered []client.Object
+		for _, wh := range warehouses.Items {
+			if wh.Name == name {
+				filtered = append(filtered, &wh)
+			}
+		}
```

Comment:
Planning ahead for supporting more resource kinds, would it be better for this filtering by name to occur after the switch instead of being repeated for each kind?